containers/payments/tokenHelper.js

```diff
@@ -0,0 +1,135 @@
+import { PAYMENT_METHOD_TYPES, PAYMENT_TOKEN_STATUS } from 'proton-shared/lib/constants';
+import { createToken, getTokenStatus } from 'proton-shared/lib/api/payments';
+import { wait } from 'proton-shared/lib/helpers/promise';
+import { c } from 'ttag';
+
+const {
+    STATUS_PENDING,
+    STATUS_CHARGEABLE,
+    STATUS_FAILED,
+    STATUS_CONSUMED,
+    STATUS_NOT_SUPPORTED
+} = PAYMENT_TOKEN_STATUS;
+
+const DELAY_PULLING = 5000;
+const DELAY_LISTENING = 1000;
+
+const pull = async ({ timer = 0, Token, api }) => {
+    if (timer > DELAY_PULLING * 30) {
+        throw new Error(c('Error').t`Payment process cancelled`);
+    }
+
+    const { Status } = await api(getTokenStatus(Token));
+
+    if (Status === STATUS_FAILED) {
+        throw new Error(c('Error').t`Payment process failed`);
+    }
+
+    if (Status === STATUS_CONSUMED) {
+        throw new Error(c('Error').t`Payment process consumed`);
+    }
+
+    if (Status === STATUS_NOT_SUPPORTED) {
+        throw new Error(c('Error').t`Payment process not supported`);
+    }
+
+    if (Status === STATUS_CHARGEABLE) {
+        return;
+    }
+
+    if (Status === STATUS_PENDING) {
+        await wait(DELAY_PULLING);
+        return pull({ Token, api, timer: timer + DELAY_PULLING });
+    }
+
+    throw new Error(c('Error').t`Unknown payment token status`);
+};
+
+const process = ({ ApprovalURL, Token, api }) => {
+    return new Promise((resolve, reject) => {
+        let listen = false;
+        const tab = window.open(ApprovalURL, '3DS support');
+
+        const reset = () => {
+            window.removeEventListener('message', onMessage, false);
+            tab.close();
+        };
+
+        const listenTab = async () => {
+            if (!listen) {
+                return;
+            }
+
+            if (tab.closed) {
+                throw new Error(c('Error').t`Tab closed`);
+            }
+
+            await wait(DELAY_LISTENING);
+            return listenTab(tab);
+        };
+
+        const onMessage = (event) => {
+            const origin = event.origin || event.originalEvent.origin; // For Chrome, the origin property is in the event.originalEvent object.
+
+            if (origin !== 'https://secure.protonmail.com') {
+                return reject();
+            }
+
+            if (event.source !== tab) {
+                return reject();
```

Comment:
I don't think we need to `reject()` here or at the origin check

Reply:
But if I don't `reject` here, who take care of the promise process? I don't like to count on `listenTab` for it.

Reply:
Oh I understand your point now. You are right.